kernel/src/fs/rootfs.rs

```diff
@@ -14,17 +15,42 @@ use super::{
 };
 use crate::prelude::*;
 
+struct BoxedReader<'a>(Box<dyn Read + 'a>);
+
+impl<'a> BoxedReader<'a> {
+    pub fn new(reader: Box<dyn Read + 'a>) -> Self {
+        BoxedReader(reader)
+    }
+}
+
+impl Read for BoxedReader<'_> {
+    fn read(&mut self, buf: &mut [u8]) -> core2::io::Result<usize> {
+        self.0.read(buf)
+    }
+}
+
 /// Unpack and prepare the rootfs from the initramfs CPIO buffer.
 pub fn init(initramfs_buf: &[u8]) -> Result<()> {
     init_root_mount();
     procfs::init();
 
-    println!("[kernel] unpacking the initramfs.cpio.gz to rootfs ...");
-    let fs = FsResolver::new();
-    let mut decoder = CpioDecoder::new(
-        GZipDecoder::new(initramfs_buf)
-            .map_err(|_| Error::with_message(Errno::EINVAL, "invalid gzip buffer"))?,
+    let mut initramfs_suffix = "";
+    let reader = match &initramfs_buf[..4] {
+        // Gzip magic number: 0x1F 0x8B
+        &[0x1F, 0x8B, _, _] => {
+            initramfs_suffix = ".gz";
+            let gzip_decoder = GZipDecoder::new(initramfs_buf)
+                .map_err(|_| Error::with_message(Errno::EINVAL, "invalid gzip buffer"))?;
+            BoxedReader::new(Box::new(gzip_decoder))
+        }
+        _ => BoxedReader::new(Box::new(Cursor::new(initramfs_buf))),
+    };
+    println!(
+        "[kernel] unpacking the initramfs.cpio{} to rootfs ...",
+        initramfs_suffix
     );
+    let mut decoder = CpioDecoder::new(reader);
```

Comment:
To improve readability (slightly), refactor the code this way by hiding `initramfs_suffix`.

```rust
    let reader = {
        let mut initramfs_suffix = "";
        let reader = match &initramfs_buf[..4] {
             ...
        };

        println!(
            "[kernel] unpacking the initramfs.cpio{} to rootfs ...",
            initramfs_suffix
        );

        reader
    });
    let mut decoder = CpioDecoder::new(reader);
```